topi/tests/python/test_fifo_buffer.py

```diff
@@ -0,0 +1,200 @@
+            s = topi.generic.schedule_injective([out])
+            update_input_window = tvm.build(s, [inc_input, input_window, out], device,
+                                            name='update_input_window')
+
+            out = topi.nn.conv2d(input_window, kernel, strides=stride, padding=padding,
+                                 dilation=dilate, layout='NCHW', out_dtype=dtype)
+            s = topi.generic.schedule_conv2d_nchw([out])
+            conv2d = tvm.build(s, [input_window, kernel, out], device, name='conv2d')
+
+        input_window_tvm = tvm.nd.array(input_window_np, ctx=ctx)
+        new_input_window_tvm = tvm.nd.empty(shape=input_window_shape, ctx=ctx, dtype=dtype)
+        kernel_tvm = tvm.nd.array(kernel_np, ctx=ctx)
+        context_tvm = tvm.nd.array(context_np, ctx=ctx)
+        new_context_tvm = tvm.nd.empty(shape=context_shape, ctx=ctx, dtype=dtype)
+        inc_output_tvm = tvm.nd.empty(shape=inc_output_shape, ctx=ctx, dtype=dtype)
+        output_window_tvm = tvm.nd.array(output_window_np, ctx=ctx)
+        new_output_window_tvm = tvm.nd.empty(shape=output_window_shape, ctx=ctx, dtype=dtype)
+        output_window_ref_tvm = tvm.nd.empty(shape=output_window_shape, ctx=ctx, dtype=dtype)
+
+        for i in range(num_iteration):
+            # Take i-th slice of inc_input_np
+            inc_input_tvm = tvm.nd.array(inc_input_np[i], ctx=ctx)
+
+            # Compute new output window incrementally, using the FIFO buffer op
+            update_context(inc_input_tvm, context_tvm, new_context_tvm)
+            conv2d_inc(new_context_tvm, kernel_tvm, inc_output_tvm)
+            update_output_window(inc_output_tvm, output_window_tvm, new_output_window_tvm)
+            context_tvm = new_context_tvm
+            output_window_tvm = new_output_window_tvm
+
+            # Compute full input window, so that we have a baseline
+            update_input_window(inc_input_tvm, input_window_tvm, new_input_window_tvm)
+            input_window_tvm = new_input_window_tvm
+            conv2d(input_window_tvm, kernel_tvm, output_window_ref_tvm)
+            # Incrementally updating the output window should be equivalent to computing it from
+            # scratch using the input window
+            tvm.testing.assert_allclose(output_window_tvm.asnumpy(),
+                                        output_window_ref_tvm.asnumpy())
+
+    for device in get_all_backend():
+        check_device(device)
+
+def test_fifo_buffer():
+    for ndim in [1, 2, 3, 4, 5, 6]:
+        for axis in range(ndim):
+            buffer_shape = tuple(7 for _ in range(ndim))
+            data_shape = tuple((2 if i == axis else 7) for i in range(ndim))
+            print('Testing FIFO buffer op: buffer_shape = {}, data_shape = {}, axis = {}'
+                  .format(buffer_shape, data_shape, axis))
+            verify_fifo_buffer(buffer_shape, data_shape, axis)
+
+def test_conv2d_integration():
+    print('Testing FIFO buffer with 1D convolution')
```

Comment:
Should it be 2D convolution? I think we probably don't need these prints.

Reply:
It's 1D because I set height to be 1.

Reply:
Actually, I shouldn't be using `conv2d` in the method name then. Let me fix it.